{
  "cells": [
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "title: \"Getting Started\"\n",
        "subtitle: \"week 01\"\n",
        "\n",
        "format:\n",
        "  html:\n",
        "    code-fold: true\n",
        "    code-tools: FALSE\n",
        "    code-copy: true\n",
        "    highlight-style: github\n",
        "---"
      ],
      "id": "77dd5557"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Text Reading\n",
        "\n",
        "How to Learn Bioinformatics 1-18;\\\n",
        "Setting Up and Managing a Bioinformatics Project 21-35;\\\n",
        "Remedial Unix Shell 37-54\n",
        "\n",
        "# Expectations\n",
        "\n",
        "# Setting up for Success!\n",
        "\n",
        "As part of this class you will be learning fundamental skills in working with genomic data. In addition you will be carrying out an independent project throughout the quarter. Generally Tuesday will be learning a skillset and Thursday will be working on your independent project.\n",
        "\n",
        "Each student will have two GitHub repositories, one where you complete \"classwork\" and as one devoted to your project. Both need to be in the organization `course-fish546-2023`.\n",
        "\n",
        "The name of these repos:\\\n",
        "preferredname-*classwork* and\\\n",
        "preferredname-projectdescriptor\n",
        "\n",
        "::: callout-important\n",
        "Make sure you have you local repo clone in logical location (eg \\~/Documents/GitHub) and that you do not move, nor place in Dropbox or similar syncing directory.\n",
        ":::\n",
        "\n",
        "Be sure to comply with guidelines\n",
        "\n",
        "::: panel-tabset\n",
        "## File Structure\n",
        "\n",
        "-   Good file structure\n",
        "\n",
        "    -   All project files in one main folder\n",
        "\n",
        "    -   Subfolders (data, code, output)\n",
        "\n",
        "-   Main folder is R project\n",
        "\n",
        "    -   Self-contained project\n",
        "\n",
        "    -   Use relative instead of absolute paths\n",
        "\n",
        "-   Good folder & file names\n",
        "\n",
        "    -   Descriptive but not too long\n",
        "\n",
        "    -   No spaces\n",
        "\n",
        "    -   Consistent format\n",
        "\n",
        "## Data\n",
        "\n",
        "-   Raw data\n",
        "\n",
        "    -   In separate folder from cleaned data\n",
        "\n",
        "    -   Never change!\n",
        "\n",
        "    -   Each file should have metadata\n",
        "\n",
        "## Code\n",
        "\n",
        "-   Scripts with code\n",
        "\n",
        "    -   Relative file paths to read in and create files\n",
        "\n",
        "    -   Lots of comments\n",
        "\n",
        "    -   Order: libraries, data, user-created functions, everything else\n",
        "\n",
        "    -   Good variable & column names\n",
        ":::\n",
        "\n",
        "::: callout-warning\n",
        "Max GitHub file size is 100MB\n",
        ":::\n",
        "\n",
        "There will be times when files are two big to include in repositories (or even you laptop). There will also be times when you have to run code outside of a GitHub repository. You will need to determine a way to effectively document this in your repository.\n",
        "\n",
        "# Working in the command-line\n",
        "\n",
        "Having already reviewed the prep material and completed the [bash tutorial](https://sr320.github.io/course-fish546-2023/assignments/00-bash.html) you are now ready to get to coding.\n",
        "\n",
        "For the first task you will take an unknown multi-fasta file and annotate it using blast. You are welcome to do this in terminal, Rstudio, or jupyter. My recommendation, and how I will demonstrate is using Rmarkdown. Once you have have your project structured, we will download software, databases, a fasta file and run the code.\n",
        "\n",
        "In your code directory create a file.\n",
        "\n",
        "`01-blast.Rmd`\n",
        "\n",
        "::: callout-tip\n",
        "Rmarkdown is a good option as you can use markdown, add pictures and more!\n",
        ":::\n",
        "\n",
        "## Downloading software\n",
        "\n",
        "NCBI Blast Software is at\n",
        "\n",
        "<https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/>\n",
        "\n",
        "::: callout-tip\n",
        "It is best to decide on central location on computer where software will be downloaded.\n",
        ":::\n",
        "\n",
        "https://github.com/RobertsLab/code/blob/master/09-blast.ipynb\n",
        "\n",
        "    cd /Applications/bioinfo/\n",
        "    curl -O https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/ncbi-blast-2.13.0+-x64-macosx.tar.gz\n",
        "    tar -xf ncbi-blast-2.13.0+-x64-macosx.tar.gz\n",
        "    cd -\n",
        "\n",
        "```\n",
        "ls /Applications/bioinfo/\n",
        "\n",
        "```\n",
        "\n",
        "```\n",
        "/Applications/bioinfo/ncbi-blast-2.13.0+/bin/blastx -h\n",
        "```\n",
        "\n",
        "## Make Blast Database\n",
        "\n",
        "see <https://www.uniprot.org/downloads>\n",
        "\n",
        "https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz\n",
        "\n",
        "\n",
        "```{cd ../data}\n",
        "curl -O https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz\n",
        "mv uniprot_sprot.fasta.gz uniprot_sprot_r2023_01.fasta.gz\n",
        "gunzip -k uniprot_sprot_r2023_01.fasta.gz\n",
        "ls ../data\n",
        "```"
      ],
      "id": "5dcaf2d8"
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}